File: frontend/src/lib/privacy.ts
import { apiRequest } from "@/lib/api";

export type ConsentStatus = {
  required: boolean;
  accepted: boolean;
  policy_version: string;
  accepted_at: string | null;
};

export type AcceptConsentResponse = {
  accepted: boolean;
  policy_version: string;
  accepted_at: string;
};

export type RevokeConsentResponse = {
  accepted: boolean;
  policy_version: string;
  revoked_at: string;
};

export type ArchiveAccountResponse = {
  archived: boolean;
  archived_at: string;
  revoked_refresh_tokens: number;
  revoked_sharing_consents: number;
  consent_revoked: boolean;
};

export type PrivacyExport = {
  exported_at: string;
  user: Record<string, unknown>;
  consent_records: Record<string, unknown>[];
  sharing_consents_granted: Record<string, unknown>[];
  sharing_consents_received: Record<string, unknown>[];
  journal_entries: Record<string, unknown>[];
  emotion_logs: Record<string, unknown>[];
  emotional_reports: Record<string, unknown>[];
  chat_sessions: Record<string, unknown>[];
  sos_events: Record<string, unknown>[];
  care_reminders: Record<string, unknown>[];
};

export type PrivacyAuditEntry = {
  id: string;
  action: string;
  actor_user_id?: string | null;
  target_user_id?: string | null;
  resource_type: string;
  resource_id?: string | null;
  metadata?: unknown;
  created_at: string;
};

export async function getConsentStatus() {
  return apiRequest<ConsentStatus>("/privacy/consent");
}

export async function getPrivacyAudit() {
  return apiRequest<PrivacyAuditEntry[]>("/privacy/audit?limit=8");
}

export async function acceptConsent(policyVersion: string) {
  return apiRequest<AcceptConsentResponse>("/privacy/consent", {
    method: "POST",
    body: JSON.stringify({ policy_version: policyVersion })
  });
}

export async function revokeConsent() {
  return apiRequest<RevokeConsentResponse>("/privacy/consent/revoke", {
    method: "POST"
  });
}

export async function archiveAccount() {
  return apiRequest<ArchiveAccountResponse>("/privacy/account/archive", {
    method: "POST"
  });
}

export async function exportPrivacyData() {
  return apiRequest<PrivacyExport>("/privacy/export");
}

File: backend/app/schemas/privacy.py
from pydantic import BaseModel


class ConsentStatusResponse(BaseModel):
    required: bool
    accepted: bool
    policy_version: str
    accepted_at: str | None = None


class AcceptConsentRequest(BaseModel):
    policy_version: str = "2026-05-27-mvp"


class AcceptConsentResponse(BaseModel):
    accepted: bool
    policy_version: str
    accepted_at: str


class RevokeConsentResponse(BaseModel):
    accepted: bool
    policy_version: str
    revoked_at: str


class ArchiveAccountResponse(BaseModel):
    archived: bool
    archived_at: str
    revoked_refresh_tokens: int
    revoked_sharing_consents: int
    consent_revoked: bool


class PrivacyExportResponse(BaseModel):
    exported_at: str
    user: dict[str, object]
    consent_records: list[dict[str, object]]
    sharing_consents_granted: list[dict[str, object]]
    sharing_consents_received: list[dict[str, object]]
    journal_entries: list[dict[str, object]]
    emotion_logs: list[dict[str, object]]
    emotional_reports: list[dict[str, object]]
    chat_sessions: list[dict[str, object]]
    sos_events: list[dict[str, object]]
    care_reminders: list[dict[str, object]]


class PrivacyAuditEntry(BaseModel):
    id: str
    action: str
    actor_user_id: str | None = None
    target_user_id: str | None = None
    resource_type: str
    resource_id: str | None = None
    metadata: dict[str, object] | list[object] | str | None = None
    created_at: str
